### PR TITLE
Shared Load Balancer

### LINKED ISSUE #7: OSL Central Load Balancer
From our meeting we came to the conclusion to using multiple data bags for managing projects within haproxy. Within each databag you can pass in an array for the frontends, listeners, and backends which will get merged with all of the other project data bags in the chef run. Here is an example of the project databag.

```json
{
  "id": "mozillazine",
  "cluster": "osl",
  "frontend": {
    "http_https": [
      "acl mozillazine hdr(host) mozillazine.org",
      "use-backend mozillazine if mozillazine"
    ],
    "http": [
      "http-request redirect scheme https if !{ ssl_fc }",
     ],
  },
  "backend": {
    "mozillazine": [
      "server mozillazine-www1 140.x.x.x:80 weight 1 maxconn 100 check",
      "server mozillazine-www2 140.x.x.x:80 weight 1 maxconn 100 check"
     ]
  }
}
```

For the backends we have the option of using a default config for backends like we do with phpbb and instead pass individual attributes to override. If we decide to take this route it will mean more logic within the recipe that may have to be rewritten when we need a custom backend.

The frontends will be created with the lwrp in the recipe since we need them to pass in the merged arrays.

I also have the idea of adding a config check before we do the final service reload with:

`haproxy -f /etc/haproxy/haproxy.cfg -c`

### PR DESCRIPTION
**Pull request by armiller**
This still needs tests but I wanted to get some feedback. Will resolve #7

### CODE REVIEW

File: metadata.rb [hunk 1]
```diff
@@ -10,3 +10,5 @@
 depends          'haproxy', '>= 1.6.6'
 depends          'osl-munin'
 depends          'rsyslog'
+depends          'certificate'
+depends          'resource_from_hash'
```
Comment (jordane) [inline metadata.rb]: It makes me happy that this actually turned out to be useful.

File: recipes/shared.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+# set up global config hashes
+backend = {}
+tmp_backends = {}
+frontend = {}
+frontend['http'] = []
+frontend['https'] = []
+crt_names = []
+
+config.each do |project|
+  p_config = data_bag_item('osl-haproxy', project)
```
Comment (jordane) [inline recipes/shared.rb]: This should be encased in a begin-rescue block.

File: recipes/shared.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  # Collect frontends
+  if p_config['frontend']
```
Comment (jordane) [inline recipes/shared.rb]: I prefer the following here:

```ruby
if p_config['frontend'].is_a? Enumerable
  ...
end
```

File: recipes/shared.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    p_config['frontend'].each_pair do |listeners, cfg|
+      if listeners == 'https_http'
```
Comment (jordane) [inline recipes/shared.rb]: `case` would make this look cleaner.

File: recipes/shared.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+node.default['haproxy']['enable_ssl'] = false
+node.default['haproxy']['enable_default_http'] = false
+node.default['haproxy']['ssl_member_port'] = 443
+
+config = data_bag('osl-haproxy')
+
+# set up global config hashes
+backend = {}
+tmp_backends = {}
```
Comment (jordane) [inline recipes/shared.rb]: Why is this needed? A comment here would be good.

File: recipes/shared.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+frontend = {}
```
Comment (jordane) [inline recipes/shared.rb]: Given what is below,

```ruby
frontend = Hash.new([])
```

might be more appropriate here.

File: recipes/shared.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    end
+  end
+
+  # Collect backends
+  if p_config['backend']
+    # Make sure backends are in the correct config
+    p_config['backend'].each do |name, attr|
+      unless attr.is_a?(Array)
+        Chef::Log.fatal("Backend server #{name} must be an Array of hashes")
+        fail
```
Comment (jordane) [inline recipes/shared.rb]: It's best to pass fail a useful error message it can exit with.

File: recipes/shared.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+config.each do |project|
+  p_config = data_bag_item('osl-haproxy', project)
+
+  # Collect frontends
+  if p_config['frontend']
+    p_config['frontend'].each_pair do |listeners, cfg|
+      if listeners == 'https_http'
+        frontend['https'] += cfg
+        frontend['http'] += cfg
+      elsif listeners == 'https'
```
Comment (jordane) [inline recipes/shared.rb]: You could take care of all the other cases with

```ruby
else
    frontend[listeners] += cfg
```

File: recipes/shared.rb [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+frontend = Hash.new([])
+frontend['http'] = []
+frontend['https'] = []
+crt_names = []
+
+config.each do |project|
+  begin
+    p_config = data_bag_item('osl-haproxy', project)
+  rescue
+    next
```
Comment (armiller) [inline recipes/shared.rb]: I don't think this is the best option. Should we just fail?

Comment (jordane) [inline recipes/shared.rb]: Probably. Make sure to pass the error message on though.

File: test/integration/data_bags/osl-haproxy/project1.json [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  "frontend": {
+    "https_http": [
+      "acl project1_http hdr(host) project1.example.com",
+      "use_backend servers-backend1 if project1_http"
+    ]
+  },
+  "backend": {
+    "backend1": [{
+      "hostname": "backend1",
+      "ipaddress": "1.1.1.1",
```
Comment (jordane) [inline test/integration/data_bags/osl-haproxy/project1.json]: There are ip addresses set aside for use in random testing like this. We should use one of those instead.

Comment (armiller) [inline test/integration/data_bags/osl-haproxy/project1.json]: Where do I find those?

Comment (jordane) [inline test/integration/data_bags/osl-haproxy/project1.json]: [Here](http://en.wikipedia.org/wiki/Reserved_IP_addresses)

**Comment by ramereth** [comment]
This might be beyond the scope of this pull request, but how do I define more listeners? Example being, I'd like to use HAProxy as a frontend for most of the OpenStack API's which require a few other ports. Might make sense to add that support in another PR after this gets merged. But it would be nice to know how that logic might be added.

**Comment by armiller** [comment]
I think we originally decided to only deal with http and https for the listeners. I can add the ability to pass in custom listeners.

You would just pass in a `listeners` key.
``` json
{
  "id": "project",
  "listeners": [
  ],
  "frontend": [
  ],
  "backend: [
  ]
}
```

**Comment by ramereth** [comment]
@armiller I think it's going to be important to have the ability to add custom listener for cases such as OpenStack. Having default http/https is fine IMO. Mind adding a test case for that?

**Comment by armiller** [comment]
@ramereth could you give me an example config?

**Comment by armiller** [comment]
@osuosl-cookbooks/chefs I added a readme and listener support. some review would be nice. If you have any ideas on how to tests this better please let me know.

File: README.md [hunk 11]
```diff
@@ -1,45 +1,57 @@
 osl-haproxy Cookbook
 ====================
 
-Site specific OSL haproxy cookbook.
+OSL haproxy wrapper cookbook. Installs HAProxy 1.5 stable from source. Default recipes are available to setup basic http and https frontends. This cookbook also contains a recipe for managing OSL's central load balancer through data bags.
```
Comment (ramereth) [inline README.md]: Line wrap.

File: README.md [hunk 12]
```diff
@@ -1,45 +1,57 @@
...
 
-Attributes
-----------
-TODO: List you cookbook attributes here.
+Recipes
+=======
+
+default
+-------
+Installs HAProxy 1.5 from source and enables the default http and optional ssl frontends. Backend servers are passed in the `['haproxy']['members']` attribute as a array of hashes.
```
Comment (ramereth) [inline README.md]: Line wrap

File: README.md [hunk 13]
```diff
@@ -1,45 +1,57 @@
...
+
+shared
+------
+This recipe manages the central OSL load balancer. It contains logic for managing the haproxy config through data bags. There is a `base` databag which handles the basic config needed for haproxy. Project specific configuration goes inside a project databag. This databag must be within the `osl-haproxy` folder. Example data bags can be viewed in `tests/integration/data_bags`.
```
Comment (ramereth) [inline README.md]: Line wrap

File: test/integration/data_bags/osl-haproxy/base.json [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  "id": "base",
+  "frontend": {
+    "http": [
+      "bind 0.0.0.0:80"
+    ],
+    "https_http": [
+      "maxconn 2000",
+      "option forwardfor"
+    ],
+    "test_tcp": [
```
Comment (ramereth) [inline test/integration/data_bags/osl-haproxy/base.json]: Would it make sense to move this into its own data bag setting up the frontend and also creating its own backend for testing which is also using the tcp mode?

Comment (armiller) [inline test/integration/data_bags/osl-haproxy/base.json]: We can just add the backend in `base.json`. If its not a project config I would rather it stay in base so we keep it simple. Base config / project config.

File: recipes/shared.rb [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+
+if node['osl-haproxy']['cluster'].nil?
+  Chef::Log.fatal("No cluster defined in node['osl-haproxy']['cluster']")
+  fail 'Error no cluster defined'
+end
+
+config.each do |project|
+  p_config = data_bag_item('osl-haproxy', project)
+
+  next unless p_config['cluster'] == node['osl-haproxy']['cluster']
```
Comment (ramereth) [inline recipes/shared.rb]: I don't think we want to use the same data bag for all clusters do we? I was thinking we could have a different data bag per cluster unless @jordane thinks that's a crazy idea.

Comment (jordane) [inline recipes/shared.rb]: As discussed in meatspace, I +1 this idea.

**Comment by armiller** [comment]
I will push the change tomorrow.

**Comment by armiller** [comment]
@osuosl-cookbooks/chefs alright. I think this is ready. Would like some review.

File: recipes/data_bag.rb [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+      end
+    end
+  end
+
+  # Collect backends
+  if p_config['backend'].is_a? Enumerable
+    # Make sure backends are in the correct config
+    p_config['backend'].each do |name, attr|
+      unless attr.is_a?(Array)
+        Chef::Log.fatal("Backend server #{name} must be an Array of hashes")
```
Comment (jordane) [inline recipes/data_bag.rb]: It would be useful to log what type of object `attr` is.

File: recipes/data_bag.rb [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+        frontend['https'] += cfg
+        frontend['http'] += cfg
+      else
+        frontend[listen] += cfg
+      end
+    end
+  end
+
+  # Collect backends
+  if p_config['backend'].is_a? Enumerable
```
Comment (jordane) [inline recipes/data_bag.rb]: Why is there no else case for this if statement?

File: recipes/data_bag.rb (hunk 16 above)
Comment (armiller) [inline recipes/data_bag.rb]: Good idea.

File: recipes/data_bag.rb (hunk 17 above)
Comment (armiller) [inline recipes/data_bag.rb]: Should there be? I don't see a reason to have one. Could you explain in more detail?

File: recipes/data_bag.rb [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+node.default['haproxy']['enable_ssl'] = false
+node.default['haproxy']['enable_default_http'] = false
+node.default['haproxy']['ssl_member_port'] = 443
+
+if node['osl-haproxy']['cluster'].nil?
```
Comment (jordane) [inline recipes/data_bag.rb]: You also should probably check for `.empty?`

File: recipes/data_bag.rb (hunk 17 above)
Comment (jordane) [inline recipes/data_bag.rb]: Sure. I'm curious why you're bothering to check if it's an `Enumerable`, when you don't really care if it isn't. It seems to me if it isn't, that it should throw a warn log. Otherwise it will fail to add the backend correctly, and fail silently.

Comment (armiller) [inline recipes/data_bag.rb]: I see. Maybe instead of `if node['foo']['bar'].is_a? Enumerable` Just have `if node['foo']['bar']`? I am just checking if that hash exists and if so adding it to the collection. I think you originally mentioned checking for Enumerable instead of just if the key exists or not. I personally am fine with `if node['foo']['bar']`. Another option would just be to log out to chef that there are no backends/frontends/listeners. Would look like:
```Ruby
if node['foo']['bar']
  # Add to collection
else
  Chef.log "foo data bag contains no bars"
end
```

File: recipes/data_bag.rb (hunk 18 above)
Comment (armiller) [inline recipes/data_bag.rb]: Good point. I will add that check.

File: recipes/data_bag.rb (hunk 17 above)
Comment (jordane) [inline recipes/data_bag.rb]: Hmm. It might be best to check if

* `p_config['backends']` is `Enumerable`
* `p_config['backends'].empty?` is `false`

And then log if either of those fail.

Comment (armiller) [inline recipes/data_bag.rb]: Sounds good. What log level? I am thinking `warn`.

Comment (jordane) [inline recipes/data_bag.rb]: `warn` works for me.

**Comment by armiller** [comment]
@jordane I would hold off on review for now. I need to make some changes.

**Comment by armiller** [comment]
My most recent commit lints the data bag and will fail with a useful message if it determines an issue. This adds a lot more logic which would be nice to test with chefspec. Granted this would take about another week or so for me to write.

**Comment by jordane** [comment]
@armiller You can abstract the linting to a method in `libraries/` and just use plain-old RSpec to test it. It'd be quicker and easier.

**Comment by armiller** [comment]
@jordane I was thinking about moving the lint methods to `libraries/`. Being able to test with RSpec makes a lot of sense. I will go ahead and work on that.

**Comment by armiller** [comment]
@jordane so here is a basic linter. I would like some feedback before writing tests. I feel like there is a better way to structure it. I went with a module method for readability in the data_bag.rb. I feel:
```
Linter.check_type(foo)
```
is easier to read than:
```
lint = Linter.new
lint.check_type(foo)
```

File: libraries/linter.rb [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# Basic linter for checking data_bag syntax.
+class Linter
```
Comment (jordane) [inline libraries/linter.rb]: This is a pretty generic name. How about `OSLHAProxyLinter`?

File: libraries/linter.rb [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+  def self.check_type(item, type)
```
Comment (jordane) [inline libraries/linter.rb]: no need for `self.`

File: libraries/linter.rb (hunk 19 above)
Comment (armiller) [inline libraries/linter.rb]: hmmm.. ok. In that case then I will just switch to standard instance methods. `OSLHAProxyLinter.check_type(foo)` is going to look ugly.

File: libraries/linter.rb [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+    unless item.is_a?(type)
+      error(
+        "Wrong type. You passed a #{item.class.name}. " \
+        "Please pass a #{type}"
+      )
+    end
+    error("You can not pass an empty #{type}.") if item.empty?
+  end
+
+  def self.check_values(item)
```
Comment (jordane) [inline libraries/linter.rb]: This should be a less generic name, since it is expecting a specific value structure.

File: libraries/linter.rb (hunk 20 above)
Comment (armiller) [inline libraries/linter.rb]: This was needed for module methods calls IE `Linter.check_type(foo)` instead of:
```
lint = Linter.new
lint.check_type(foo)
```
I am going to switch this. (see comment above).

Comment (jordane) [inline libraries/linter.rb]: You may also want to do a `check_ancestry`. Chef is prone to inheriting things just to rename them. [`Chef::DelayedEvaluator`](https://github.com/chef/chef/blob/master/lib/chef/mixin/params_validate.rb#L19-L20) is one example.

File: libraries/linter.rb [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+    check_type(item, Array)
+    item.each do |cfg|
+      if !cfg.is_a?(String) || cfg.empty?
+        error('Haproxy config values must be a String.')
+      end
+    end
+  end
+
+  def self.error(message)
+    Chef::Log.fatal(message)
```
Comment (jordane) [inline libraries/linter.rb]: If you want to test this independently, you'll also want to `require 'chef/log'`

File: libraries/linter.rb (hunk 19 above)
Comment (jordane) [inline libraries/linter.rb]: Seems reasonable.

File: test/integration/databag/serverspec/server_spec.rb [hunk 23]
```diff
@@ -0,0 +1,43 @@
+require 'serverspec'
+
+set :backend, :exec
+
+describe file('/etc/haproxy/haproxy.cfg') do
+  its(:content) { should match(/bind 0.0.0.0:80/) }
+  its(:content) { should match(/maxconn 2000/) }
+  its(:content) { should match(/option forwardfor/) }
+  its(:content) do
+    should match(%r{bind 0.0.0.0:443 crt \/etc\/pki\/tls\/project1-www.pem \
```
Comment (jordane) [inline test/integration/databag/serverspec/server_spec.rb]: Isn't one of the points of using `%r{}` not having to escape `/`?

**Comment by armiller** [comment]
@jordane so I had it working just fine until I tried to get the spec and libraries files to work with rubocop. Now I am getting a name error:
```
Failure/Error: expect do
       expected Exception with "Error in data bag syntax", got #<NameError: uninitialized constant OSLHAProxyLinter::Chef> with backtrace:
         # ./libraries/osl_haproxy_linter.rb:43:in `error'
         # ./libraries/osl_haproxy_linter.rb:28:in `check_type'
         # ./spec/linter_spec.rb:11:in `block (4 levels) in <top (required)>'
         # ./spec/linter_spec.rb:10:in `block (3 levels) in <top (required)>'
         # ./spec/linter_spec.rb:8:in `each'
         # ./spec/linter_spec.rb:8:in `block (2 levels) in <top (required)>'
     # ./spec/linter_spec.rb:10:in `block (3 levels) in <top (required)>'
     # ./spec/linter_spec.rb:8:in `each'
     # ./spec/linter_spec.rb:8:in `block (2 levels) in <top (required)>'
```
Looks to be a problem with the namespace. What should it inherit from?

**Comment by jordane** [comment]
@armiller It's a bit of a deceiving error that is the cause of 2 problems that combined together.

Both problems occur on line 43 of `libraries/osl_haproxy_linter`. The first is that `Chef::Log` is actually a relatively namespaced class, and Ruby looks for `OSLHAProxyLinter::Chef::Log`, but can't find `OSLHAProxyLinter::Chef`. You can fix this by prepending `::` to make it an absolute namespace: `::Chef::Log`.

The second problem is that you need to use `require` before you actually pull in any of the Chef classes. You can fix this by either putting the require in the library, or putting it in the spec file. I recommend the latter.

Note that you will also have to include 'Chef' in the Gemfile in order for 'bundle exec rspec' to work properly.

**Comment by armiller** [comment]
@jordane Thanks for the suggestion. I went ahead with mocking `Chef::Log`. That way we don't have to include the large `Chef` gem.

File: spec/linter_spec.rb [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+        'compare' => 'Array',
+        'value' => { 'foo' => 'bar' }
+      },
+      {
+        'compare' => 'Hash',
+        'value' => %w(testing1 testing2)
+      }
+    ]
+    bad_types.each do |test|
+      ob = Object.const_get(test['compare']).new
```
Comment (jordane) [inline spec/linter_spec.rb]: You can simplify this by doing `'compare' => Array,` above, and then just doing `ob = test['compare'].new` here.

File: spec/linter_spec.rb [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  it 'should not raise error with a valid type' do
+    linter = OSLHAProxyLinter.new
+    good_checks = {
+      'Hash' => { 'foo' => 'bar' },
+      'Array' => %w(testing1 testing2)
+    }
+    good_checks.each_pair do |type, values|
+      ob = Object.const_get(type).new
```
Comment (jordane) [inline spec/linter_spec.rb]: You can do the same here. For better or worse, Ruby lets you use any object as a hash key. so `Hash => { 'foo' => 'bar' }`

File: spec/linter_spec.rb [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+      expect do
+        linter.check_type(values, ob.class)
```
Comment (jordane) [inline spec/linter_spec.rb]: Taking into account my comments above, this would become `linter.check_type(values, type)`

File: spec/linter_spec.rb (hunk 24 above)
Comment (armiller) [inline spec/linter_spec.rb]: oh cool. didn't realize that.

File: spec/linter_spec.rb [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+  end
+end
+
+describe 'OSLHAProxyLinter get_type method' do
+  it 'should raise error with an empty type' do
+    linter = OSLHAProxyLinter.new
+    empty_check = [Hash, Array]
+    empty_check.each do |type|
+      expect do
+        linter.check_type(type.new, type.class)
```
Comment (jordane) [inline spec/linter_spec.rb]: This isn't quite right:

```
irb(main):001:0> Array
=> Array
irb(main):002:0> Array.class
=> Class
irb(main):003:0> type = Array
=> Array
irb(main):004:0> type.class
=> Class
```

**Comment by armiller** [comment]
@osuosl-cookbooks/chefs I think this pr is in a pretty good place. Some final review would be nice. If you have any ideas in to how we can improve our test coverage, please let me know.

File: README.md [hunk 28]
```diff
@@ -1,45 +1,67 @@
 osl-haproxy Cookbook
 ====================
 
-Site specific OSL haproxy cookbook.
+OSL haproxy wrapper cookbook. Installs HAProxy 1.5 stable from source.
```
Comment (ramereth) [inline README.md]: This needs to be updated once we merge the package install back.

File: README.md [hunk 29]
```diff
@@ -1,45 +1,67 @@
...
 
-Attributes
-----------
-TODO: List you cookbook attributes here.
+Recipes
+=======
+
+default
+-------
+Installs HAProxy 1.5 from source and enables the default http and optional ssl
```
Comment (ramereth) [inline README.md]: Will be installed via package..

File: recipes/data_bag.rb [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+      crt_names << crt['attributes']['cert_file']
+      resource_from_hash crt['name'] do
+        hash crt
+        action :do
+      end
+    end
+  end
+end
+
+# This has been copied from the haproxy cookbook
```
Comment (ramereth) [inline recipes/data_bag.rb]: Since this cookbook is a little more complicated, it might be helpful to add a few more comments below about what exactly is going on. Doesn't have to go into much detail, but at least state the summary of what its doing. That will make it easier to troubleshoot later. It can be as short as you did above.

**Comment by ramereth** [comment]
@armiller I'm not sure if is a rare edge case, but I considered a situation where we might want to use this data bag method for non http/https load balancing (only), thus not needing http/https frontends. (I seem to recall we talked about this at once, so I apologize if we're re-hashing this).

I can at least see a situation where we may not want https for simple load balancing needs. Right now if you remove all the references to http, https, or http_https in the data bags, it still leaves an empty frontends defined in the config. Ideally the code should check to see if the array is empty, and if so not to build that frontend.

I'm just concerned about having http/https enabled by default all the time even if you don't have them defined.

**Comment by jordane** [comment]
@ramereth did you actually run the code and generate an empty list of frontends, or are you worried that iterating over an empty list will create empty frontends? `[].each {|i| # do stuff }` runs exactly 0 times.

**Comment by ramereth** [comment]
@jordane I created a new suite locally, copied the data bags and created a new "cluster". I removed all of the bits about http/https/http_https, created a new frontend called keystone and made it a tcp frontend and then converged.  It generated the frontends like this:

```
frontend http

frontend https
  bind 0.0.0.0:443

frontend keystone
  acl project1_http hdr(host) project1.example.com
  acl project2_http hdr(host) project2.example.org
  bind 0.0.0.0:8000
  mode tcp
  use_backend servers-backend1 if project1_http
  use_backend servers-backend2 if project2_http
```

If you'd like, I can upload what data bags I used exactly.

**Comment by jordane** [comment]
@ramereth Cool, I just wanted to make sure before I dove in. I'm writing up a couple more test cases that we need to handle.

**Comment by jordane** [comment]
@armiller I pushed up some failing test suites to cover @ramereth's use case.

**Comment by armiller** [comment]
Ah yes. That should be an easy fix. I will get to that later this afternoon.

**Comment by armiller** [comment]
@ramereth should be good now. @jordane thanks for setting up those suites.

**Comment by ramereth** [comment]
The only thing we don't test is if we don't have either http/https but I tested it myself and it does work. not sure if we want to add that test or not. Otherwise I give my :+1:

**Comment by armiller** [comment]
@osuosl-cookbooks/chefs I have rebased the branch. This is ready for some final review.

**Comment by ramereth** [comment]
:+1:
